From 28effec8dbc91da8e30f42d11f1d96dc636db47f Mon Sep 17 00:00:00 2001
From: Samuel Dowling <samuel.dowling@protonmail.com>
Date: Thu, 30 Sep 2021 23:51:35 +0930
Subject: [PATCH] Guard dependency discovery

* Add guards to prevent usage of custom cmake find package scripts.
---
 CMakeLists.txt | 81 ++++++++++++++++++++++++++++++++++++++++++--------
 1 file changed, 69 insertions(+), 12 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7857f8c..31cdb8e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -274,7 +274,13 @@ if(APPLE)
   set(ARMA_USE_ACCELERATE true)
   
   if(ALLOW_OPENBLAS_MACOS)
-    include(ARMA_FindOpenBLAS)
+    if(USE_SYSTEM_OPENBLAS)
+        include(ARMA_FindOpenBLAS)
+    elseif(USE_OPENBLAS)
+        find_package(OpenBLAS)
+    else()
+        set(OpenBLAS_FOUND NO)
+    endif()
     message(STATUS "OpenBLAS_FOUND = ${OpenBLAS_FOUND}")
     message(STATUS "")
     message(STATUS "*** If use of OpenBLAS is causing problems,")
@@ -289,8 +295,16 @@ if(APPLE)
   endif()
   
   if(ALLOW_BLAS_LAPACK_MACOS)
-    include(ARMA_FindBLAS)
-    include(ARMA_FindLAPACK)
+    if(USE_SYSTEM_BLAS)
+      include(ARMA_FindBLAS)
+    else()
+      set(BLAS_FOUND NO)
+    endif()
+    if(USE_SYSTEM_LAPACK)
+      include(ARMA_FindLAPACK)
+    else()
+      set(LAPACK_FOUND NO)
+    endif()
     message(STATUS "  BLAS_FOUND = ${BLAS_FOUND}"  )
     message(STATUS "LAPACK_FOUND = ${LAPACK_FOUND}")
     message(STATUS "")
@@ -329,14 +343,44 @@ if(APPLE)
   
 else()
   
-  include(ARMA_FindMKL)
-  include(ARMA_FindOpenBLAS)
-  include(ARMA_FindATLAS)
-  include(ARMA_FindBLAS)
-  include(ARMA_FindLAPACK)
+  if(USE_SYSTEM_MKL)
+    include(ARMA_FindMKL)
+  else()
+    set(MKL_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_OPENBLAS)
+    include(ARMA_FindOpenBLAS)
+  elseif(USE_OPENBLAS)
+    find_package(OpenBLAS)
+  else()
+    set(OpenBLAS_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_ATLAS)
+    include(ARMA_FindATLAS)
+  else()
+    set(ATLAS_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_BLAS)
+    include(ARMA_FindBLAS)
+  else()
+    set(BLAS_FOUND NO)
+  endif()
+
+  if(USE_SYSTEM_LAPACK)
+    include(ARMA_FindLAPACK)
+  else()
+    set(LAPACK_FOUND NO)
+  endif()
   
   if(ALLOW_FLEXIBLAS_LINUX AND (${CMAKE_SYSTEM_NAME} MATCHES "Linux"))
-    include(ARMA_FindFlexiBLAS)
+    if(USE_SYSTEM_FLEXIBLAS)
+      include(ARMA_FindFlexiBLAS)
+    else()
+      set(FlexiBLAS_FOUND NO)
+    endif()
   endif()
   
   message(STATUS "      MKL_FOUND = ${MKL_FOUND}"       )
@@ -470,7 +514,12 @@ if(DETECT_HDF5)
     # HDF5_INCLUDE_DIRS is the correct include directory.  So, in either case we
     # can use the first element in the list.  Issue a status message, too, just
     # for good measure.
-    list(GET HDF5_INCLUDE_DIRS 0 ARMA_HDF5_INCLUDE_DIR)
+    if(NOT USE_SYSTEM_HDF5)
+      list(GET HDF5_INCLUDE_DIRS 1 ARMA_HDF5_INCLUDE_DIR)
+    else()
+      list(GET HDF5_INCLUDE_DIRS 0 ARMA_HDF5_INCLUDE_DIR)
+    endif()
+
     message(STATUS "ARMA_HDF5_INCLUDE_DIR = ${ARMA_HDF5_INCLUDE_DIR}")
     message(STATUS "")
     message(STATUS "*** If use of HDF5 is causing problems,")
@@ -480,7 +529,11 @@ if(DETECT_HDF5)
   endif()
 endif()
 
-include(ARMA_FindARPACK)
+if(USE_SYSTEM_ARPACK)
+  include(ARMA_FindARPACK)
+else()
+  set(ARPACK_FOUND NO)
+endif()
 message(STATUS "ARPACK_FOUND = ${ARPACK_FOUND}")
 
 if(ARPACK_FOUND)
@@ -488,7 +541,11 @@ if(ARPACK_FOUND)
   set(ARMA_LIBS ${ARMA_LIBS} ${ARPACK_LIBRARY})
 endif()
 
-include(ARMA_FindSuperLU5)
+if(USE_SYSTEM_SUPERLU)
+  include(ARMA_FindSuperLU5)
+else()
+  set(SuperLU_FOUND NO)
+endif()
 message(STATUS "SuperLU_FOUND = ${SuperLU_FOUND}")
 
 if(SuperLU_FOUND)
-- 
2.33.0

